### PR TITLE
CPDP-1588: dms submission rate limiter

### PR DESCRIPTION
**Pull request by mikail-khan**

### CODE REVIEW

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        if (workItem.failureCount === failureCountLimit) {
+          val _ = dmsSubmissionService.setResultStatus(workItem.id, PermanentlyFailed)
+          ()
+        } else {
+          logger.info(
+            s"DMS Submission poller: " +
+              s"processing work item ${workItem.toString}"
+          )
+
+          implicit val hc: HeaderCarrier = workItem.item.headerCarrier
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:66]: will the bearer token still be valid after a certain amount of time? Isn't this the same problem we encountered when we tried to do retries on failed tax enrolments outside of someone's session?

File: conf/application.conf [hunk 2]
```diff
@@ -237,6 +240,23 @@ email {
 dms {
     queue-name = "queue-name"
     b64-business-area = "YnVzaW5lc3MtYXJlYQ=="
+    submission-poller {
+        initial-delay = 60 seconds
+        interval = 60 seconds
+        failure-count-limit = 50
+        in-progress-retry-after = 5000
+        mongo {
+            ttl = 604800 seconds # 7 days
```
Comment (AndyHWChung) [inline conf/application.conf:249]: couldn't you just write 7 days?

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala (hunk 1 above)
Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:66]: it doesn't look like it's authenticated actually, maybe we're ok: 
https://github.com/hmrc/gform/blob/master/app/uk/gov/hmrc/gform/dms/DmsSubmissionController.scala#L40

File: app/uk/gov/hmrc/cgtpropertydisposals/controllers/returns/SubmitReturnsController.scala [hunk 3]
```diff
@@ -64,16 +65,20 @@ class SubmitReturnsController @Inject() (
...
+            _                  <- dmsSubmissionService.enqueue(
+                                    DmsSubmissionRequest(
+                                      request.headerCarrier,
+                                      sanitisedHtml,
+                                      submissionResult.formBundleId,
+                                      returnRequest.subscribedDetails.cgtReference,
+                                      returnRequest.completeReturn
+                                    )
+                                  )
+            _                   = logger.info(
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/controllers/returns/SubmitReturnsController.scala:77]: do you mean change the name of `enqueue` to what you suggest or change the name of the type `DmsSubmissionRequest`?

[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposals/controllers/returns/SubmitReturnsController.scala:77]: I meant in the log message

File: app/uk/gov/hmrc/cgtpropertydisposals/repositories/dms/DmsSubmissionRepo.scala [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    new WorkItemFieldNames {
+      val receivedAt   = "receivedAt"
+      val updatedAt    = "updatedAt"
+      val availableAt  = "availableAt"
+      val status       = "status"
+      val id           = "_id"
+      val failureCount = "failureCount"
+    }
+
+  override def inProgressRetryAfterProperty: String = "dms.submission-poller.in-progress-retry-after"
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/repositories/dms/DmsSubmissionRepo.scala:75]: `work-item-repo` is badly coded, has poor abstractions, and the config is all over the place - we have no choice here.

[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/repositories/dms/DmsSubmissionRepo.scala:75]: oh yeah, and true to form, it's totally undocumented!! :angry:

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala (hunk 1 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:66]: it should be ok, we just really require the bearer token when making the `gforms` request - wasn't very enthused about storing the header-carrier away but in reality that is what we are doing - suspending the request until some future point in time.

File: app/uk/gov/hmrc/cgtpropertydisposals/repositories/dms/DmsSubmissionRepo.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposals/repositories/dms/DmsSubmissionRepo.scala:75]: I noticed that when I was trying to look for info in their README :worried:

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+class DmsSubmissionPoller @Inject() (
+  actorSystem: ActorSystem,
+  dmsSubmissionService: DmsSubmissionService,
+  dmsSubmissionPollerContext: DmsSubmissionPollerContext,
+  servicesConfig: ServicesConfig
+)(implicit
+  executionContext: ExecutionContext
+) extends Logging {
+
+  private val initialPollerDelay: FiniteDuration =
```
Comment (mikail-khan) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:43]: add jitter here

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                logger.warn(
+                  s"DMS Submission poller: " +
+                    s"submission failed for work item ${workItem.toString} " +
+                    s"with error - re-schedule for retry: ${error.toString}"
+                )
+                val _ = dmsSubmissionService.setProcessingStatus(workItem.id, Failed)
+              case Right(envelopeId) =>
+                logger.info(
+                  s"DMS Submission poller: " +
+                    s"submission succeeded for work item ${workItem.toString} " +
```
Comment (mikail-khan) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:87]: only return the basic info required that enable investigation

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+              workItem.item.formBundleId,
+              workItem.item.cgtReference,
+              workItem.item.completeReturn
+            )
+            .value
+            .map {
+              case Left(error)       =>
+                logger.warn(
+                  s"DMS Submission poller: " +
+                    s"submission failed for work item ${workItem.toString} " +
```
Comment (mikail-khan) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:80]: same here

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionRequest.scala [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+final case class DmsSubmissionRequest(
+  headerCarrier: HeaderCarrier,
+  html: B64Html,
+  formBundleId: String,
+  cgtReference: CgtReference,
+  completeReturn: CompleteReturn
+)
+
+object DmsSubmissionRequest {
+  implicit val authorizationFormat: OFormat[Authorization]               = Json.format
```
Comment (mikail-khan) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionRequest.scala:36]: take out headercarrier and test

File: conf/application.conf [hunk 9]
```diff
@@ -237,6 +240,23 @@ email {
 dms {
     queue-name = "queue-name"
     b64-business-area = "YnVzaW5lc3MtYXJlYQ=="
+    submission-poller {
+        initial-delay = 60 seconds
+        interval = 60 seconds
```
Comment (mikail-khan) [inline conf/application.conf:245]: set interval to `120`, in-progress-retry-after to double the interval value

File: conf/application.conf [hunk 10]
```diff
@@ -237,6 +240,23 @@ email {
...
+    }
+}
+
+dms-submission-poller-dispatcher {
+  fork-join-executor {
+    parallelism-factor = 1
+    parallelism-max = 1
+  }
+  throughput = 100
 }
```
Comment (mikail-khan) [inline conf/application.conf:260]: andy to investigate

File: app/uk/gov/hmrc/cgtpropertydisposals/controllers/returns/SubmitReturnsController.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/controllers/returns/SubmitReturnsController.scala:77]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  private val failureCountLimit: Int = servicesConfig.getInt("dms.submission-poller.failure-count-limit")
+
+  val _ = actorSystem.scheduler.schedule(initialPollerDelay, pollerInterval)(poller())(dmsSubmissionPollerContext)
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  def poller(): Unit = {
+    val _ = dmsSubmissionService.dequeue.map {
+      case Some(workItem) =>
+        if (workItem.failureCount === failureCountLimit) {
+          val _ = dmsSubmissionService.setResultStatus(workItem.id, PermanentlyFailed)
```
[the remark this answers is not in the document]
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:76]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala (hunk 1 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:81]: hmm...I tried to take the header carrier out which meant removing the implicit that is threaded across the modules which are used to make submissions and I hit an issue with the generation of the temporary file created to make the multipart form request. The session-id in the `hc` is used to as part of the file name in order to ensure that it is unqiue. That can be changed but it is hitting quite a few methods so not a small change. I say we go with this and create a tech debt ticket to resolve this later.

File: conf/application.conf (hunk 2 above)
Comment (mikail-khan) [reply conf/application.conf]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala (hunk 5 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala]: done

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala (hunk 6 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala]: fixed

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala (hunk 7 above)
Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala]: fixed

File: conf/application.conf (hunk 9 above)
Comment (mikail-khan) [reply conf/application.conf]: done

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala (hunk 1 above)
Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:81]: ok, let's leave it as it is then  - we can address that later if we need to. It's easier to take stuff out than it is to put things in so we should be good

File: app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+)(implicit
+  executionContext: DmsSubmissionPollerExecutionContext
+) extends Logging {
+
+  private val jitteredInitialDelay: FiniteDuration = FiniteDuration(
+    servicesConfig.getDuration("dms.submission-poller.initial-delay").toNanos,
+    TimeUnit.NANOSECONDS
+  ) + FiniteDuration(
+    Random.nextInt((servicesConfig.getDuration("dms.submission-poller.jitter-period").toSeconds.toInt + 1)).toLong,
+    TimeUnit.NANOSECONDS
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:55]: shouldn't this be `TimeUnit.SECONDS`?

Comment (AndyHWChung) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala:55]: or should the `toSeconds` be `toNanos` above?

Comment (mikail-khan) [reply app/uk/gov/hmrc/cgtpropertydisposals/service/dms/DmsSubmissionPoller.scala]: fixed - time units should be SECONDS.

**Comment by AndyHWChung** [comment]
LGTM

**Pull request merged**